File: py/DuelingGeneratorsAPI.py
import queue

MOD_DIVISOR = 2147483647
A_FACTOR = 16807
B_FACTOR = 48271

def generateNext(prevValue, factor, modDivisor =  MOD_DIVISOR):
	return (prevValue * factor) % modDivisor

def int2bin(val):
	return '{0:0>16b}'.format(val)[-16:]

def judgeMatch(val1, val2):
	bin1 = int2bin(val1)
	bin2 = int2bin(val2)
	return bin1 == bin2

def judgeNMatches(init1, init2, n):
	val1 = init1
	val2 = init2
	count = 0
	for i in range(n):
		val1 = generateNext(val1, A_FACTOR, MOD_DIVISOR)
		val2 = generateNext(val2, B_FACTOR, MOD_DIVISOR)
		if judgeMatch(val1, val2):
			count = count + 1
	return count

def judgeNMatchesPickyGenerators(init1, init2, n):
	val1 = init1
	val2 = init2
	count = 0
	q1 = queue.Queue();
	q2 = queue.Queue();
	i = 0
	while i < n:
		val1 = generateNext(val1, A_FACTOR, MOD_DIVISOR)
		if (val1 % 4) == 0:
			q1.put(val1)
		val2 = generateNext(val2, B_FACTOR, MOD_DIVISOR)
		if (val2 % 8) == 0:
			q2.put(val2)
		if not q1.empty() and not q2.empty():
			i = i+1;
			if judgeMatch(q1.get(), q2.get()):
				count = count + 1
	return count


print('test output is: ' + str(judgeNMatches(65, 8921, 5)))
#print('real output is: ' + str(judgeNMatches(634, 301, 40000000)))

print('test 2 output is: ' + str(judgeNMatchesPickyGenerators(65, 8921, 1056)))
print('real output 2 is: ' + str(judgeNMatchesPickyGenerators(634, 301, 5000000)))

File: js/DuelingGeneratorsAPI.js
var DuelingGeneratorsAPI = {};

const MOD_DIVISOR = 2147483647;
const GEN_A_FACTOR = 16807;
const GEN_B_FACTOR = 48271;

function padString(str, width, filler) {
    return (filler.repeat(width) + str).substr(-width);
}

DuelingGeneratorsAPI.generator = function* generator(init, factor, modDivisor) {
    var val = init;
    while(true) {
        val = (val * factor) % modDivisor;
        var rt = yield val;
        if (rt != undefined) {
            val = rt;
        }
    }
};

DuelingGeneratorsAPI.judge = function(val1, val2) {
    var bin1 = padString(val1.value.toString(2), 16, '0');
    var bin2 = padString(val2.value.toString(2), 16, '0');
    return bin1 == bin2;
};


DuelingGeneratorsAPI.judgeGenerators = function(init1, init2, iterations) {
    var gen1 = DuelingGeneratorsAPI.generator(init1, GEN_A_FACTOR, MOD_DIVISOR);
    var gen2 = DuelingGeneratorsAPI.generator(init2, GEN_B_FACTOR, MOD_DIVISOR);
    var count = 0;
    for (var i=0; i<iterations; i++) {
        if (DuelingGeneratorsAPI.judge(gen1.next(), gen2.next())) {
            count += 1;
        }
    }
    return count;
};

DuelingGeneratorsAPI.judgeMatchesIn40MilGenerations = function(initialValuesString) {
    vals = initialValuesString.split(',');
    return DuelingGeneratorsAPI.judgeGenerators(parseInt(vals[0]), parseInt(vals[1]), 40000000);
};

//TESTING
var testCount = DuelingGeneratorsAPI.judgeGenerators(65, 8921, 5);
if (testCount != 1) throw new Error("testCount == " + testCount);
var realCount = DuelingGeneratorsAPI.judgeGenerators(634, 301, 4000000);
if (realCount != 1) throw new Error("realCount == " + realCount);
else console.log("realCount == " + realCount);